### DISCUSSION TITLE
Change Model Via the API

### DISCUSSION THREAD

**Discussion by mckinlec** [oobabooga/text-generation-webui#1713] [Q&A]
Is it possible to change the model via the API? Also, is there an API docs page? I can't seem to find it, I don't have an API link at the bottom of my UI

**Comment by Dynathresh** [comment]
You can post a request to load a different model.
```
req = {
        'action': 'load',
        'model_name': model,
}  # You can tack on more params to this request - see the reference link at bottom of comment
response = requests.post(f'http://{HOST}/api/v1/**model**', json=request)  # This will swap out the model running on the server
return response.json()  # This just shows you the server response to you loading the new model, to help you verify it loaded
```
I haven't yet found a way to ask the server which models are available.

All learned from here: https://github.com/oobabooga/text-generation-webui/blob/4ba30f676544600c5c9ffdddfc50bfb4682f8a36/api-examples/api-example-model.py#L42

**Reply by mckinlec** [reply]
Thanks, that's been added since I last took a look, I started using KoboldAI instead but that has issues putting the model on the GPU when loading through the api. I'll take another look at this api

**Comment by matatonic** [comment]
See here for how listing models is used:

https://github.com/oobabooga/text-generation-webui/blob/4ba30f676544600c5c9ffdddfc50bfb4682f8a36/api-examples/api-example-model.py#L132

Ex:
```py
def model_api(request):
    response = requests.post(f'http://{HOST}/api/v1/model', json=request)
    return response.json()

all_models = model_api({'action': 'list'})['result']

for m in all_models:
   ...
```

**Reply by Dynathresh** [reply]
Perfect, this is very useful to me. I appreciate it. Is there documentation on top of the api-examples or a way to learn all the possible requests like this?

**Reply by matatonic** [reply]
Read the source code for the blocking API in the api extension, that's probably the best way at the moment. There may be better docs some day, but not yet. There is also an #api-dev-help channel in the official discord.

**Reply by deevis** [reply]
As a matter of interest, setting the model via 

```
model_api({'action': 'load', 'model_name': 'my_favorite_model'})
```

works fine for the API but will _leave the webUI in an inconsistent state_ with the old model still displaying on the `Model` tab - refreshing the webUI page does not reconcile this...

**Reply by dsfcoll** [reply]
Is this response is still relevant? I get 404 for POST /api/v1/model

**Reply by bradylangdale** [reply]
@VladStepanov It's changed a little bit. Here's the [implementation](https://github.com/oobabooga/text-generation-webui/blob/ad1ff53034b97db4dcbeeae3cbf5e2e7fdca0297/extensions/openai/script.py#L280-L307) for reference. 

You can call it like this.
```python
def load_model(base_url, args_json={}):
    response = requests.post(base_url + 'internal/model/load/', json=args_json)
    return response.json()

url = 'http://localhost:5000/v1/'
args = { "model_name": "Llama2" }

load_model(url, args)
```

**Reply by dsfcoll** [reply]
Thanks! Works perfect

**Reply by SalomonKisters** [reply]
Hey guys, this doesnt work anymore, does anyone know how to do it now?

**Reply by Grego3j** [reply]
@SalomonKisters Do you have the API command-line flag enabled? Checking these boxes helped me: 
![Screenshot from 2024-03-24 18-56-42](https://github.com/oobabooga/text-generation-webui/assets/95664698/ec2ef6c2-2f49-427b-a12e-f96767ad5092)

**Reply by SalomonKisters** [reply]
Yes, I do. This is my command:
bash -c '/app/modules/text-generation-webui/start_linux.sh --listen --listen-host 0.0.0.0 --model $MODEL_FOLDER_NAME --api --api-key $DEFAULT_API_KEY --nowebui'

**Comment by deevis** [comment]
For those arriving here late to the game, and using API calls...

In your CMD_FLAGS.txt ( or checked in Session => Boolean Command Line Flags )
```
# make the api available on port 5000
--api 
# enable the openai extension
--extensions openai
# start the web interface
--listen
```

Once running, these API calls allow you to check/list/modify models...

_To check the currently loaded model_:  
`GET /v1/internal/model/info`   ( then `response.json()['model_name']` )

_To list available models_: 
`GET /v1/internal/model/list`     ( then `response.json()['model_names']` )

_To load a different model_: 
`POST /v1/internal/model/load` ( payload: `{'model_name': model_name}`)

**Comment by bradylangdale** [comment]
@SalomonKisters anyone who comes across this thread.

Here's an [example](https://github.com/bradylangdale/LLM_World/blob/master/oobabooga.py) from a small project you use for reference.

Be sure to have the flags enabled that are mentioned above. 

```python3
import requests

class Oobabooga:

    def __init__(self, ip='localhost', port='5000') -> None:
        self.ip = ip
        self.port = port

        self.url = f'http://{self.ip}:{self.port}/v1/'

    def get_models(self):
        response = requests.get(self.url + 'internal/model/list')
        return response.json()['model_names']
    
    def load_model(self, args_json={}):
        response = requests.post(self.url + 'internal/model/load/', json=args_json)
        return response.json()
    
    def unload_model(self, args_json={}):
        response = requests.post(self.url + 'internal/model/unload/', json=args_json)
        return response.json()
    
    def get_model_info(self):
        response = requests.get(self.url + 'internal/model/info')
        return response.json()
    
    def stop_generation(self):
        response = requests.post(self.url + 'internal/stop-generation')
        return response.json()
    
    def chat_completion(self, args_json={}, stream=False):
        if not stream:
            response = requests.post(self.url + 'chat/completions', json=args_json)
            return response.json()
        else:
            return requests.post(self.url + 'chat/completions',
                                 headers={ "Content-Type": "application/json" },
                                 json=args_json,
                                 verify=False,
                                 stream=True)
    
    def completion(self, args_json={}, stream=False):
        if not stream:
            response = requests.post(self.url + 'completions', json=args_json)
            return response.json()
        else:
            return requests.post(self.url + 'completions',
                                 headers={ "Content-Type": "application/json" },
                                 json=args_json,
                                 verify=False,
                                 stream=True)
```
